abci/server/socket_server.go

```diff
@@ -67,33 +69,40 @@ func (s *SocketServer) OnStop() {
 	s.connsMtx.Lock()
 	defer s.connsMtx.Unlock()
 
-	for id, conn := range s.conns {
-		delete(s.conns, id)
-		if err := conn.Close(); err != nil {
-			s.logger.Error("error closing connection", "id", id, "conn", conn, "err", err)
+	for id := range s.conns {
+		if err := s.unsafeRmConn(id); err != nil {
+			s.logger.Error("error closing connection", "id", id, "err", err)
 		}
 	}
 }
 
-func (s *SocketServer) addConn(conn net.Conn) int {
+func (s *SocketServer) addConn(conn net.Conn, closer func()) int {
 	s.connsMtx.Lock()
 	defer s.connsMtx.Unlock()
 
 	connID := s.nextConnID
 	s.nextConnID++
 	s.conns[connID] = conn
-
+	s.connsClose[connID] = closer
 	return connID
 }
 
 // deletes conn even if close errs
 func (s *SocketServer) rmConn(connID int) error {
 	s.connsMtx.Lock()
 	defer s.connsMtx.Unlock()
+	return s.unsafeRmConn(connID)
+}
+
+func (s *SocketServer) unsafeRmConn(connID int) error {
+	if closer, ok := s.connsClose[connID]; ok {
```

Comment:
Given that we're already under the lock here, and `net.Conn` implements `io.Closer`, do we need this separate map? It appears if we find a connection in `s.conns` we're already dropping it and closing it.

I see various closer functions plumbed around as parameters, but I maybe have missed where they're originating. Are these "close a related thing" someplace?

Reply:
`unsafeRmConn` does three things:
- cleans up the map(s)
- call the cancel function for the context that the connection handling threads are running for each connection
- closes the actual connection . 

This is more confusing than it needs to be and I think I can make something work that's more simple.